internal/provider/alert_resource.go

```diff
@@ -0,0 +1,273 @@
+	getIdFunc         idGetterFunc[T]
+	getPipelineIdFunc idGetterFunc[T]
+	schema            schema.Schema
+}
+
+func (r *AlertResource[T]) TypeName() string {
+	// example: mezmo_threshold_alert
+	return fmt.Sprintf("%s_%s_alert", PROVIDER_TYPE_NAME, r.typeName)
+}
+
+func (r *AlertResource[T]) NodeType() string {
+	// example: threshold_alert
+	return fmt.Sprintf("%s_alert", r.typeName)
+}
+
+func (r *AlertResource[T]) TerraformSchema() schema.Schema {
+	return r.schema
+}
+
+func (r *AlertResource[T]) ConvertToTerraformModel(component *reflect.Value) (*reflect.Value, error) {
+	if !component.CanInterface() {
+		return nil, errors.New("alert Value does not contain an interfaceable type")
+	}
+
+	alert, ok := component.Interface().(Alert)
+	if !ok {
+		return nil, errors.New("component Value cannot be cast to an Alert")
+	}
+
+	var model T
+	r.toModelFunc(&model, &alert)
+	res := reflect.ValueOf(model)
+	return &res, nil
+}
+
+// Configure implements resource.ResourceWithConfigure.
+func (r *AlertResource[T]) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected client.Client, got: %T. Please report this issue to Mezmo.", req.ProviderData),
+		)
+		return
+	}
+
+	r.client = client
+}
+
+// Create implements resource.Resource.
+func (r *AlertResource[T]) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Retrieve values from plan
+	var plan T
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	component, dd := r.fromModelFunc(&plan, nil)
+	if setDiagnosticsHasError(dd, &resp.Diagnostics) {
+		return
+	}
+
+	if os.Getenv("DEBUG_ALERT") == "1" {
+		fmt.Printf("----- Plan ----- %+v\n", plan)
+		PrintJSON("----- Alert TO Create api ---", component)
+	}
+
+	stored, err := r.client.CreateAlert(r.getPipelineIdFunc(&plan).ValueString(), component)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating alert",
+			"Could not create alert, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	if os.Getenv("DEBUG_ALERT") == "1" {
+		PrintJSON("----- Alert FROM Create api ---", stored)
+	}
+
+	NullifyPlanFields(&plan, r.schema)
+
+	r.toModelFunc(&plan, stored)
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+}
+
+// Delete implements resource.Resource.
+func (r *AlertResource[T]) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state T
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// For this delete, we need to fetch the Alert model so that we can access
+	// `component_kind` and `component_id`. The values for "plan" and "previousState"
```

Comment:
I'm a little confused by the comment. `component_kind` is a simple enum with the value of `source`, `transform`, `sink`, and now `alert`. Don't we just know what the kind is based on what resource triggered the call to this method?

Reply:
ah, this `component_kind` is use in our REST routes as a path param.  That discussion about those routes and how to form their paths is what prompts this comment.  We decided on `/v3/pipeline/:pipeline_id/:component_kind/:component_id/alert` as a basic format, so here, `component_kind` is the kind of alert.  So, it's an enum of `source` `transform` for now.

There's no real way to know that value unless we dug in the TF state to try and pull out values for `component_id`.  Since `component_id` is also needed, we already have the object to pull out the kind.  The comment was supposed to say how we need the full alert struct, and not just the `id` as normal deletes do.  I'll clean it up.